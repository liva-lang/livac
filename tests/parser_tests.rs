use insta::assert_snapshot;
use livac::{lexer::tokenize, parser::parse};
use std::fs;

/// Test helper para casos correctos del parser
fn test_parser_ok(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/ok_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: ok_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let ast = parse(tokens, &source).unwrap();
    
    // Convertir AST a JSON para snapshot
    let json = serde_json::to_string_pretty(&ast).unwrap();
    assert_snapshot!(format!("ok_{}.ast", test_name), json);
}

/// Test helper para casos de error del parser
fn test_parser_err(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/err_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: err_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let result = parse(tokens, &source);
    assert!(result.is_err(), "Expected parser error for test: {}", test_name);
    
    let error_msg = result.unwrap_err().to_string();
    assert_snapshot!(format!("err_{}.diag", test_name), error_msg);
}

#[test]
fn test_functions_oneliner() {
    test_parser_ok("functions_oneliner");
}

#[test]
fn test_functions_block() {
    test_parser_ok("functions_block");
}

#[test]
fn test_classes() {
    test_parser_ok("classes");
}

#[test]
fn test_control_flow() {
    test_parser_ok("control_flow");
}

#[test]
fn test_expressions() {
    test_parser_ok("expressions");
}

#[test]
fn test_imports() {
    test_parser_ok("imports");
}

#[test]
fn test_unclosed_paren() {
    test_parser_err("unclosed_paren");
}

#[test]
fn test_unclosed_brace() {
    test_parser_err("unclosed_brace");
}

#[test]
fn test_case_without_switch() {
    test_parser_err("case_without_switch");
}

#[test]
fn test_return_outside_function() {
    test_parser_err("return_outside_function");
}

#[test]
fn test_duplicate_default() {
    test_parser_err("duplicate_default");
}
