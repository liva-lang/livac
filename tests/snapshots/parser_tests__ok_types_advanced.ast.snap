---
source: tests/parser_tests.rs
expression: json
---
{
  "items": [
    {
      "Function": {
        "name": "main",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "i8_val",
                    "type_ref": {
                      "Simple": "i8"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 127
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "i16_val",
                    "type_ref": {
                      "Simple": "i16"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 32767
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "i32_val",
                    "type_ref": {
                      "Simple": "i32"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 2147483647
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "i64_val",
                    "type_ref": {
                      "Simple": "i64"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 9223372036854775807
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "i128_val",
                    "type_ref": {
                      "Simple": "i128"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 1000000000000000000
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Identifier": "i128"
                }
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "u8_val",
                    "type_ref": {
                      "Simple": "u8"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 255
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "u16_val",
                    "type_ref": {
                      "Simple": "u16"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 65535
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "u32_val",
                    "type_ref": {
                      "Simple": "u32"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 4294967295
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Identifier": "u32"
                }
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "u64_val",
                    "type_ref": {
                      "Simple": "u64"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 1000000000000
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Identifier": "u64"
                }
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "u128_val",
                    "type_ref": {
                      "Simple": "u128"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 1000000000000000000
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Identifier": "u128"
                }
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "isize_val",
                    "type_ref": {
                      "Simple": "isize"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 1000000000000
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "usize_val",
                    "type_ref": {
                      "Simple": "usize"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 1000000000000
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "f32_val",
                    "type_ref": {
                      "Simple": "f32"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Float": 3.14
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Identifier": "f32"
                }
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "f64_val",
                    "type_ref": {
                      "Simple": "f64"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Float": 2.718281828
                  }
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Identifier": "f64"
                }
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "char_a",
                    "type_ref": {
                      "Simple": "char"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Char": "A"
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "char_b",
                    "type_ref": {
                      "Simple": "char"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Char": "B"
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "char_z",
                    "type_ref": {
                      "Simple": "char"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Char": "Z"
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "char_space",
                    "type_ref": {
                      "Simple": "char"
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Char": " "
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "bytes_val",
                    "type_ref": {
                      "Simple": "bytes"
                    }
                  }
                ],
                "init": {
                  "ArrayLiteral": [
                    {
                      "Literal": {
                        "Int": 72
                      }
                    },
                    {
                      "Literal": {
                        "Int": 101
                      }
                    },
                    {
                      "Literal": {
                        "Int": 108
                      }
                    },
                    {
                      "Literal": {
                        "Int": 108
                      }
                    },
                    {
                      "Literal": {
                        "Int": 111
                      }
                    }
                  ]
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "optional",
                    "type_ref": {
                      "Generic": {
                        "base": "Option",
                        "args": [
                          {
                            "Simple": "string"
                          }
                        ]
                      }
                    }
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "Some"
                    },
                    "args": [
                      {
                        "Literal": {
                          "String": "hello"
                        }
                      }
                    ],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "optional_none",
                    "type_ref": {
                      "Generic": {
                        "base": "Option",
                        "args": [
                          {
                            "Simple": "number"
                          }
                        ]
                      }
                    }
                  }
                ],
                "init": {
                  "Identifier": "None"
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "result_ok",
                    "type_ref": {
                      "Generic": {
                        "base": "Result",
                        "args": [
                          {
                            "Simple": "number"
                          },
                          {
                            "Simple": "string"
                          }
                        ]
                      }
                    }
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "Ok"
                    },
                    "args": [
                      {
                        "Literal": {
                          "Int": 42
                        }
                      }
                    ],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "result_err",
                    "type_ref": {
                      "Generic": {
                        "base": "Result",
                        "args": [
                          {
                            "Simple": "string"
                          },
                          {
                            "Simple": "string"
                          }
                        ]
                      }
                    }
                  }
                ],
                "init": {
                  "Call": {
                    "callee": {
                      "Identifier": "Err"
                    },
                    "args": [
                      {
                        "Literal": {
                          "String": "error"
                        }
                      }
                    ],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "int_array",
                    "type_ref": null
                  }
                ],
                "init": {
                  "ArrayLiteral": [
                    {
                      "Literal": {
                        "Int": 1
                      }
                    },
                    {
                      "Literal": {
                        "Int": 2
                      }
                    },
                    {
                      "Literal": {
                        "Int": 3
                      }
                    },
                    {
                      "Literal": {
                        "Int": 4
                      }
                    },
                    {
                      "Literal": {
                        "Int": 5
                      }
                    }
                  ]
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "name": "str_array",
                    "type_ref": null
                  }
                ],
                "init": {
                  "ArrayLiteral": [
                    {
                      "Literal": {
                        "String": "a"
                      }
                    },
                    {
                      "Literal": {
                        "String": "b"
                      }
                    },
                    {
                      "Literal": {
                        "String": "c"
                      }
                    }
                  ]
                },
                "is_fallible": false
              }
            },
            {
              "Expr": {
                "expr": {
                  "Call": {
                    "callee": {
                      "Identifier": "print"
                    },
                    "args": [
                      {
                        "Literal": {
                          "String": "All advanced types parsed successfully"
                        }
                      }
                    ],
                    "exec_policy": "normal",
                    "type_args": []
                  }
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    }
  ]
}
